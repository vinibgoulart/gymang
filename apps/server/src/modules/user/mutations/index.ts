import UserAdd from './UserAddMutation';
import UserLogin from './UserLoginMutation';
import UserLogout from './UserLogoutMutation';

const UserMutations = {
  UserAdd,
  UserLogin,
  UserLogout,
};

export default UserMutations;
