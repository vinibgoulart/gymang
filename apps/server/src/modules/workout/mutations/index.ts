import WorkoutAdd from './WorkoutAddMutation';

const WorkoutMutations = {
  WorkoutAdd,
};

export default WorkoutMutations;
