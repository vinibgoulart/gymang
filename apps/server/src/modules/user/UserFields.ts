import { isLoggedIn } from '@gymang/core';
import {
  edgeField,
  connectionArgs,
  NullConnection,
} from '@gymang/graphql';
import { UserLoader } from '@gymang/user';
import { GraphQLNonNull } from 'graphql';

import UserType, { UserConnection } from './UserType';

export const userTypeField = (key = 'user', bypassViewerCanSee = false) => ({
  [key]: {
    type: UserType,
    resolve: async (obj, args, context) => {
      return UserLoader.load(context, obj[key], bypassViewerCanSee);
    },
  },
});

export const userEdgeField = () =>
  edgeField({
    connection: UserConnection,
    load: UserLoader.load,
    edgeName: 'userEdge',
    name: 'User',
  });

export const userConnectionField = (customResolver = null) => ({
  users: {
    type: new GraphQLNonNull(UserConnection.connectionType),
    args: {
      ...connectionArgs,
      // filters: {
      //   type: UserFilterInputType,
      // },
    },
    resolve: (obj, args, context) => {
      if (customResolver) {
        return customResolver(obj, args, context);
      }

      if (!isLoggedIn(context)) {
        return NullConnection;
      }

      return UserLoader.loadAll(context, args);
    },
  },
});

export const meField = () => ({
  me: {
    type: UserType,
    resolve: (_root, _args, context) => {
      return UserLoader.load(context, context.user?._id);
    },
  },
});
