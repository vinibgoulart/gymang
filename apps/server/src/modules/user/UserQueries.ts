import { meField, userConnectionField } from './UserFields';

const UserQueries = {
  ...meField(),
  ...userConnectionField(),
};

export default UserQueries;
