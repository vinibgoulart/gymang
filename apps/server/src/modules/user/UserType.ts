import { GraphQLContext } from '@gymang/core';
import {
  connectionDefinitions,
  nodeInterface,
  registerTypeLoader,
} from '@gymang/graphql';
import { IUser, UserLoader } from '@gymang/user';
import { GraphQLObjectType, GraphQLString } from 'graphql';
import { globalIdField } from 'graphql-relay';

const UserType = new GraphQLObjectType<IUser, GraphQLContext>({
  name: 'User',
  description: 'Represents a User',
  fields: () => ({
    id: globalIdField('User'),
    firstName: {
      type: GraphQLString,
      resolve: (user) => user.firstName,
    },
    email: {
      type: GraphQLString,
      resolve: (user) => user.email,
    },
  }),
  interfaces: () => [nodeInterface],
});

export const UserConnection = connectionDefinitions({
  name: 'User',
  nodeType: UserType,
});

registerTypeLoader(UserType, UserLoader.load);

export default UserType;
