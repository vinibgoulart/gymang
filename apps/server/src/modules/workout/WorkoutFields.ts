import { isLoggedIn } from '@gymang/core';
import {
  edgeField,
  connectionArgs,
  NullConnection,
} from '@gymang/graphql';
import { WorkoutLoader } from '@gymang/workout';
import { GraphQLNonNull } from 'graphql';

import WorkoutType, { WorkoutConnection } from './WorkoutType';

export const workoutTypeField = (
  key = 'workout',
  bypassViewerCanSee = false,
) => ({
  [key]: {
    type: WorkoutType,
    resolve: async (obj, args, context) => {
      return WorkoutLoader.load(context, obj[key], bypassViewerCanSee);
    },
  },
});

export const workoutEdgeField = () =>
  edgeField({
    connection: WorkoutConnection,
    load: WorkoutLoader.load,
    edgeName: 'workoutEdge',
    name: 'Workout',
  });

export const workoutConnectionField = (customResolver = null) => ({
  workouts: {
    type: new GraphQLNonNull(WorkoutConnection.connectionType),
    args: {
      ...connectionArgs,
      // filters: {
      //   type: WorkoutFilterInputType,
      // },
    },
    resolve: (obj, args, context) => {
      if (customResolver) {
        return customResolver(obj, args, context);
      }

      if (!isLoggedIn(context)) {
        return NullConnection;
      }

      return WorkoutLoader.loadAll(context, args);
    },
  },
});
