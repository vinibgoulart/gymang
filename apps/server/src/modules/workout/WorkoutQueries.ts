import { myWorkoutConnectionField } from './WorkoutFields';

const WorkoutQueries = {
  ...myWorkoutConnectionField(),
};

export default WorkoutQueries;
