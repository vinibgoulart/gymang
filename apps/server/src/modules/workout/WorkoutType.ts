import type { GraphQLContext } from '@gymang/core';
import {
  connectionDefinitions,
  nodeInterface,
  registerTypeLoader,
} from '@gymang/graphql';
import type { IWorkout } from '@gymang/workout';
import { WorkoutLoader } from '@gymang/workout';
import { GraphQLNonNull, GraphQLObjectType, GraphQLString } from 'graphql';
import { globalIdField } from 'graphql-relay';

import { userTypeField } from '../user/UserFields';

const WorkoutType = new GraphQLObjectType<IWorkout, GraphQLContext>({
  name: 'Workout',
  description: 'Represents a Workout',
  fields: () => ({
    id: globalIdField('Workout'),
    name: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (workout) => workout.name,
    },
    description: {
      type: GraphQLString,
      resolve: (workout) => workout.description,
    },
    ...userTypeField('createdBy'),
    ...userTypeField(),
  }),
  interfaces: () => [nodeInterface],
});

export const WorkoutConnection = connectionDefinitions({
  name: 'Workout',
  nodeType: WorkoutType,
});

registerTypeLoader(WorkoutType, WorkoutLoader.load);

export default WorkoutType;
