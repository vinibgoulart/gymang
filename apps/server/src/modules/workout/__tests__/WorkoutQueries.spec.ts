import { GRAPHQL_TYPE } from '@gymang/core';
import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  sanitizeTestObject,
} from '@gymang/testutils';
import { handleCreateUser } from '@gymang/user';
import { handleCreateWorkout } from '@gymang/workout';
import { graphql } from 'graphql';

import { getContext } from '../../../getContext';
import { schema as schemaAdmin } from '../../../graphql/schema/schema';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

const query = `
  query WorkoutQueriesSpecQuery {
    myWorkouts(first: 10) {
      edges {
        node {
          id
          name
          user
          createdBy
          description
        }
      }
    }
  }
`;

it('should get a list of workouts', async () => {
  const user = await handleCreateUser();
  await handleCreateWorkout({
    user,
    createdBy: user,
  });

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaAdmin,
    source: query,
    rootValue,
    contextValue: context,
  });

  expect(result.errors).toBeUndefined();

  const [workout] = result.data.myWorkouts.edges;

  expect(workout.node.user).toEqual(user.id);
  expect(workout.node.createdBy).toEqual(user.id);

  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});

it('should get a list of workouts in a user that not create this workout', async () => {
  const user = await handleCreateUser();
  const userWithWorkout = await handleCreateUser();

  await handleCreateWorkout({
    user: userWithWorkout._id,
    createdBy: userWithWorkout._id,
  });

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaAdmin,
    source: query,
    rootValue,
    contextValue: context,
  });

  expect(result.errors).toBeUndefined();

  const [workout] = result.data.myWorkouts.edges;

  expect(workout.node.user).toEqual(userWithWorkout.id);
  expect(workout.node.createdBy).toEqual(userWithWorkout.id);

  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});
