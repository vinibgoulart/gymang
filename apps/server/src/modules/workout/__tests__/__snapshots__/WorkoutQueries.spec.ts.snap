// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get a list of workouts 1`] = `
{
  "workouts": {
    "edges": [
      {
        "node": {
          "createdBy": "ObjectId",
          "description": "EMPTY",
          "id": "FROZEN-ID",
          "name": "Workout NaN name",
          "user": "ObjectId",
        },
      },
    ],
  },
}
`;

exports[`should get a list of workouts in a user that not create this workout 1`] = `
{
  "workouts": {
    "edges": [
      {
        "node": {
          "createdBy": "ObjectId",
          "description": "EMPTY",
          "id": "FROZEN-ID",
          "name": "Workout 1 name",
          "user": "ObjectId",
        },
      },
    ],
  },
}
`;
